### PR TITLE
Integration testing

### PR DESCRIPTION
**Pull request by cjyetman**
- depends on #184
- depends on #186
- depends on #187

### CODE REVIEW

**Comment by cjyetman** [comment]
@jacobvjk and @jdhoffa I need minimum 4 files to run through a minimal integration test run. This is how I'm making them...

```r
# raw_loanbook_1.csv (sampled loantakers from r2dii.data::abcd_demo) -----------

example_loanbook_path <- "~/Desktop/test/input/loanbooks/gr1_sample_loanbook_100_121.csv"
path <- "tests/testthat/test-data/input/loanbooks/raw_loanbook_1.csv"

readr::read_csv(example_loanbook_path ,col_types = col_types_raw) %>% 
  mutate(name_direct_loantaker = sample(r2dii.data::abcd_demo$name_company, size = n())) %>% 
  readr::write_csv(path, na = "")


# ABCD.xlsx (from r2dii.data::abcd_demo) ---------------------------------------

path <- "tests/testthat/test-data/input/ABCD.xlsx"
sheet <- "Company Indicators - PACTA Comp"

r2dii.data::abcd_demo %>% 
  mutate(ald_timestamp = "2022 Q4") %>%
  mutate(company_id = as.numeric(company_id)) %>%
  mutate(year = year + 2) %>%
  relocate(
    company_id,
    name_company,
    lei,
    is_ultimate_owner,
    sector,
    technology,
    plant_location,
    year,
    production,
    production_unit,
    emission_factor,
    emission_factor_unit,
    ald_timestamp
  ) %>% 
  list() %>% 
  `names<-`(sheet) %>% 
  write_xlsx(path = path, format_headers = FALSE)


# scenario_data_sda.csv --------------------------------------------------------
# simulate from r2dii.data::co2_intensity_scenario_demo

r2dii.data::co2_intensity_scenario_demo %>% 
  readr::write_csv(
    na = "",
    file = testthat::test_path("test-data/input/scenario_data_sda.csv")
  )


# scenario_data_tms.csv --------------------------------------------------------
# simulate from r2dii.data::scenario_demo_2020

r2dii.data::scenario_demo_2020 %>% 
  readr::write_csv(
    na = "",
    file = testthat::test_path("test-data/input/scenario_data_tms.csv")
  )


# split_company_ids.csv --------------------------------------------------------

tibble(
  company_id = sample(r2dii.data::abcd_demo$name_company, size = 20),
  name_company = sample(r2dii.data::abcd_demo$company_id, size = 20)
) %>%
  readr::write_csv(
    na = "",
    file = testthat::test_path("test-data/input/split_company_ids.csv")
  )
```

Do you see any reason why I should not include these files in this PR?

**Comment by jacobvjk** [comment]
> @jacobvjk and @jdhoffa I need minimum 4 files to run through a minimal integration test run. This is how I'm making them...
> 
> ```r
> # raw_loanbook_1.csv (sampled loantakers from r2dii.data::abcd_demo) -----------
> 
> example_loanbook_path <- "~/Desktop/test/input/loanbooks/gr1_sample_loanbook_100_121.csv"
> path <- "tests/testthat/test-data/input/loanbooks/raw_loanbook_1.csv"
> 
> readr::read_csv(example_loanbook_path ,col_types = col_types_raw) %>% 
>   mutate(name_direct_loantaker = sample(r2dii.data::abcd_demo$name_company, size = n())) %>% 
>   readr::write_csv(path, na = "")
> 
> 
> # ABCD.xlsx (from r2dii.data::abcd_demo) ---------------------------------------
> 
> path <- "tests/testthat/test-data/input/ABCD.xlsx"
> sheet <- "Company Indicators - PACTA Comp"
> 
> r2dii.data::abcd_demo %>% 
>   mutate(ald_timestamp = "2022 Q4") %>%
>   mutate(company_id = as.numeric(company_id)) %>%
>   mutate(year = year + 2) %>%
>   relocate(
>     company_id,
>     name_company,
>     lei,
>     is_ultimate_owner,
>     sector,
>     technology,
>     plant_location,
>     year,
>     production,
>     production_unit,
>     emission_factor,
>     emission_factor_unit,
>     ald_timestamp
>   ) %>% 
>   list() %>% 
>   `names<-`(sheet) %>% 
>   write_xlsx(path = path, format_headers = FALSE)
> 
> 
> # scenario_data_sda.csv --------------------------------------------------------
> # simulate from r2dii.data::co2_intensity_scenario_demo
> 
> r2dii.data::co2_intensity_scenario_demo %>% 
>   readr::write_csv(
>     na = "",
>     file = testthat::test_path("test-data/input/scenario_data_sda.csv")
>   )
> 
> 
> # scenario_data_tms.csv --------------------------------------------------------
> # simulate from r2dii.data::scenario_demo_2020
> 
> r2dii.data::scenario_demo_2020 %>% 
>   readr::write_csv(
>     na = "",
>     file = testthat::test_path("test-data/input/scenario_data_tms.csv")
>   )
> 
> 
> # split_company_ids.csv --------------------------------------------------------
> 
> tibble(
>   company_id = sample(r2dii.data::abcd_demo$name_company, size = 20),
>   name_company = sample(r2dii.data::abcd_demo$company_id, size = 20)
> ) %>%
>   readr::write_csv(
>     na = "",
>     file = testthat::test_path("test-data/input/split_company_ids.csv")
>   )
> ```
> 
> Do you see any reason why I should not include these files in this PR?

Seems okay to include to me. Just one thing: the way you sample the split_company_ids data, the name and the company id will not be related in the abcd_demo data set. That is structurally not what the split_company_ids file should be. each row should contain a valid pair of name_company and company_id. Otherwise the sector split will not be applied properly here

**Comment by cjyetman** [comment]
> Seems okay to include to me. Just one thing: the way you sample the split_company_ids data, the name and the company id will not be related in the abcd_demo data set. That is structurally not what the split_company_ids file should be. each row should contain a valid pair of name_company and company_id. Otherwise the sector split will not be applied properly here

For the moment (I think) the goal is to run through the whole process without a hard error. None of the output values are required to be contextually valid, so I think that's ok for the moment anyway.

**Comment by cjyetman** [comment]
now waiting on #186 to force r2dii.match >= 0.2.0 for R 4.1 on macOS

File: tests/testthat/test-1-prepare_abcd.R [hunk 1]
```diff
@@ -0,0 +1,29 @@
+test_that("", {
```
Comment (jdhoffa) [inline tests/testthat/test-1-prepare_abcd.R]:
```suggestion
test_that("with known input, runs without error", {
```

File: tests/testthat/test-2-match_loanbooks.R [hunk 2]
```diff
@@ -0,0 +1,25 @@
+test_that("", {
```
Comment (jdhoffa) [inline tests/testthat/test-2-match_loanbooks.R]:
```suggestion
test_that("with known input, runs without error", {
```

File: tests/testthat/test-3-prioritise_and_diagnose.R [hunk 3]
```diff
@@ -0,0 +1,55 @@
+test_that("", {
```
Comment (jdhoffa) [inline tests/testthat/test-3-prioritise_and_diagnose.R]:
```suggestion
test_that("with known input, runs without error", {
```

File: tests/testthat/test-4-analyse.R [hunk 4]
```diff
@@ -0,0 +1,34 @@
+test_that("", {
```
Comment (jdhoffa) [inline tests/testthat/test-4-analyse.R]:
```suggestion
test_that("with known input, runs without error", {
```

**Review by jdhoffa — DISMISSED** [review]
Just request that the tests gain some names. 

Also is there a reason the test file names are numbered?

**Comment by cjyetman** [comment]
> Also is there a reason the test file names are numbered?

They must run in the defined order

**Comment by jdhoffa** [comment]
MacOS check seems to fail (`green_or_brown` has been deprecated as of [r2dii.data@0.5.0](https://github.com/RMI-PACTA/r2dii.data/edit/main/NEWS.md#r2diidata-050) )

**Comment by cjyetman** [comment]
now requires r2dii.analysis >= 0.3.0 for increasing_or_decreasing (#187)

**Comment by jacobvjk** [comment]
running the tests locally seems a bit brittle. when I run them, I get this:

```
✔ |          1 | 3-prioritise_and_diagnose [2.1s]            
⠏ |          0 | 4-analyse                                   [1] "Sankey plot (by sector) cannot be generated. Skipping!"
[1] "Sankey plot (by sector and company) cannot be generated. Skipping!"
[1] "Scatter plot exposure by alignment cannot be generated. Skipping!"
Scatter plot BO/PO cannot be generated. Skipping!
Scatter plot BO/PO cannot be generated. Skipping!
Scatter plot BO/PO cannot be generated at company level. Skipping!
Scatter plot BO/PO cannot be generated at company level. Skipping!
✔ |          1 | 4-analyse 
```

is that expected? It seems like `4-analyse` is run twice, once skipping, once successfully?
Also, when I now rerun the test, I get an error in `1-prepare_abcd`, because the output directory already exists (i.e. it was not cleaned up after the first run) and the test expects no conditions. Although the condition is just the warning that the output directory already exists and will be overwritten, as per #179 

```
── Failed tests ─────────────────────────────────────────────
Failure (test-1-prepare_abcd.R:28:3): with known input, runs without error
Expected `prepare_abcd(config)` to run without any conditions.
i Actually got a <simpleWarning> with text:
  Output directory `dir_prepared_abcd` already exists. The existing directory will be removed and replaced with the output of the current run.
```

**Comment by jdhoffa** [comment]
> running the tests locally seems a bit brittle. when I run them, I get this:
> 
> ```
> ✔ |          1 | 3-prioritise_and_diagnose [2.1s]            
> ⠏ |          0 | 4-analyse                                   [1] "Sankey plot (by sector) cannot be generated. Skipping!"
> [1] "Sankey plot (by sector and company) cannot be generated. Skipping!"
> [1] "Scatter plot exposure by alignment cannot be generated. Skipping!"
> Scatter plot BO/PO cannot be generated. Skipping!
> Scatter plot BO/PO cannot be generated. Skipping!
> Scatter plot BO/PO cannot be generated at company level. Skipping!
> Scatter plot BO/PO cannot be generated at company level. Skipping!
> ✔ |          1 | 4-analyse 
> ```
> 
> is that expected? It seems like `4-analyse` is run twice, once skipping, once successfully? Also, when I now rerun the test, I get an error in `1-prepare_abcd`, because the output directory already exists (i.e. it was not cleaned up after the first run) and the test expects no conditions. Although the condition is just the warning that the output directory already exists and will be overwritten, as per #179
> 
> ```
> ── Failed tests ─────────────────────────────────────────────
> Failure (test-1-prepare_abcd.R:28:3): with known input, runs without error
> Expected `prepare_abcd(config)` to run without any conditions.
> i Actually got a <simpleWarning> with text:
>   Output directory `dir_prepared_abcd` already exists. The existing directory will be removed and replaced with the output of the current run.
> ```

I think this is likely because `testthat` is designed for unit testing (e.g. testing `prepare_abcd` etc. in isolation) rather than integration testing (testing each function in order, with the output of one function used as the input for the next). 

The test skipping is probably some race condition, where the outputs of the previous test are just not complete yet

**Comment by cjyetman** [comment]
> running the tests locally seems a bit brittle. when I run them, I get this:
> 
> ```
> ✔ |          1 | 3-prioritise_and_diagnose [2.1s]            
> ⠏ |          0 | 4-analyse                                   [1] "Sankey plot (by sector) cannot be generated. Skipping!"
> [1] "Sankey plot (by sector and company) cannot be generated. Skipping!"
> [1] "Scatter plot exposure by alignment cannot be generated. Skipping!"
> Scatter plot BO/PO cannot be generated. Skipping!
> Scatter plot BO/PO cannot be generated. Skipping!
> Scatter plot BO/PO cannot be generated at company level. Skipping!
> Scatter plot BO/PO cannot be generated at company level. Skipping!
> ✔ |          1 | 4-analyse 
> ```
> 
> is that expected? It seems like `4-analyse` is run twice, once skipping, once successfully? Also, when I now rerun the test, I get an error in `1-prepare_abcd`, because the output directory already exists (i.e. it was not cleaned up after the first run) and the test expects no conditions. Although the condition is just the warning that the output directory already exists and will be overwritten, as per #179
> 
> ```
> ── Failed tests ─────────────────────────────────────────────
> Failure (test-1-prepare_abcd.R:28:3): with known input, runs without error
> Expected `prepare_abcd(config)` to run without any conditions.
> i Actually got a <simpleWarning> with text:
>   Output directory `dir_prepared_abcd` already exists. The existing directory will be removed and replaced with the output of the current run.
> ```

1 - yes, it is expected that those "sanky" lines get output to the console... it then forces testthat to reprint the test name line so that it is the most recent thing on the console, it's *not* actually running again

2 - I couldn't think of a clean way of clearing out the test files that are created during the test because they need to be intermingled with the files that need to be created pre-testing which are included in the repo. In the automated CI, it doesn't matter because a brand new clone is always made first that will always have the pre-files but not the post-files. I would dump them in a temp directory that gets deleted when the test are done, but they need to be intermingled with the input files that are already there. I can't make them use temporary files because we're using functions that have a side effect of creating files. catch-22

**Comment by cjyetman** [comment]
🎉 it works!

File: tests/testthat/test-3-prioritise_and_diagnose.R [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        start_year = 2022L,
+        time_frame = 5L,
+        by_group = "group_id"
+      ),
+      sector_split = list(
+        apply_sector_split = FALSE,
+        sector_split_type = "equal_weights"
+      ),
+      matching = list(
+        params_match_name = list(
```
Comment (jacobvjk) [inline tests/testthat/test-3-prioritise_and_diagnose.R:23]: I don't think this section should be needed here

**Review by jacobvjk — APPROVED** [review]
nit: I think the config passed to `prioritise_and_diagnose()` could be more minimal, but
lgtm

File: tests/testthat/test-3-prioritise_and_diagnose.R (hunk 5 above)
Comment (cjyetman) [reply tests/testthat/test-3-prioritise_and_diagnose.R:23]: `run_match_prioritize()` will error if `sector_split_type` is not set

https://github.com/RMI-PACTA/pacta.multi.loanbook/blob/813fdc4537972e0fbe53762c03c4be3f763fdeed/R/run_match_prioritize.R#L58

probably should fix that... will open new issue

Comment (cjyetman) [reply tests/testthat/test-3-prioritise_and_diagnose.R:23]: #189

**Comment by cjyetman** [comment]
I don't no how to get around the requested changes that have already been addressed 🤷🏻

**Comment by jdhoffa** [comment]
> I don't no how to get around the requested changes that have already been addressed 🤷🏻

Sorry! All good, approved

File: tests/testthat/test-3-prioritise_and_diagnose.R (hunk 5 above)
Comment (jacobvjk) [reply tests/testthat/test-3-prioritise_and_diagnose.R:23]: I actually meant the `matchting` section, not `sector_split_type`

Comment (cjyetman) [reply tests/testthat/test-3-prioritise_and_diagnose.R:23]: Oops. Make a new issue and I'll look at it. I found a bunch of places where some input was required even if a part of the process was being skipped

**Pull request merged**